### PR TITLE
Tracking metablocks' rounds in the Consensus contract.

### LINKED ISSUE #92: Tracking metablocks' rounds in the Consensus contract.
Add in the `Consensus` contract, the following mappings named `metablockchains` and `metablockTips`:
```solidity
  mapping(bytes20 /* chain id */ => mapping(uint256 => MetaBlock)) public metablockchains;
  mapping(bytes20 /* chain id */ => uint256 /* metablock number */) public metablockTips;
```

Metablock's struct:
```solidity
    struct Metablock {
        bytes32 metablockHash;
        uint256 roundBlockNumber;
        MetablockRound round;
    }
```

Metablock's round:
```solidity
  enum MetablockRound {
        Undefined,
        Precommitted,
        CommitteeFormed,
        CommitteeDecided,
        Committed,            
    }
```

The mapping `metablockchains` is updated in the following function calls:
  - `precommitMetablock`
  - `formCommittee`
  - `registerCommitteeDecision`
  - `commitMetablock`

To be able to update `metablockchains` and `metablockTips` mappings one of the below proposals should be picked:

Adding the following two mappings to Consensus:
```solidity
  - mapping(address /* core */ => bytes20 /* chain id */) public cores;
  - mapping(address /* committee */ => bytes20 /* chain id */) public committees;
```

or,

`Consensus` is passing `chainId` into `core` and `committee` on creation. `Core::precommitMetablock`and `Committee::registerCommitteeDecision` specify the `chainId` on calling back. 

Optional (optimization):
  - Keep only latest n number of recent blocks, prune the older data.

### LINKED ISSUE #95: Create internal method `isValidator` in consensus
- To consistently check if the address is active validator or not
- `function isValidator(address _core, address _validator)`
- Update signature of `enterCommittee` such that it now takes core address and metachainId too,
```
function enterCommittee(
        bytes32 _metachainId,
        address _core,
        address _validator,
        address _furtherMember
    )
```
- this method should be used in `enterCommittee` and `publishEndpoint` (and others)
- requires `core.isValidator(_validator)`; requires `!reputation.isSlashed(_validator)`

### PR DESCRIPTION
**Pull request by pgev**
Fixes #92 
Fixes #95

### CODE REVIEW

File: contracts/consensus/Consensus.sol [hunk 1]
```diff
@@ -91,29 +103,28 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
     /** Coinbase split per mille */
     uint256 public coinbaseSplitPerMille;
 
-    /** Block hash of heads of Metablockchains */
-    mapping(bytes20 /* chainId */ => bytes32 /* MetablockHash */) public metablockHeaderTips;
+    /** Mapping of a metablock number to a metablock (per chain).  */
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol]: "(per metachain)"; also on L109

File: contracts/consensus/Consensus.sol [hunk 2]
```diff
@@ -91,29 +103,28 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
...
+    mapping(bytes20 => mapping(uint256 => Metablock)) public metablockchains;
 
-    /** Core statuses */
-    mapping(address /* core */ => CoreStatus /* coreStatus */) public coreStatuses;
+    /** Metablocks' heights per chain. */
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol]: `/** Metablockchains' current height of tip per metachain. */`

Comment (pgev) [reply contracts/consensus/Consensus.sol]: Done.

File: contracts/consensus/Consensus.sol (hunk 1 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol]: Done.

File: contracts/consensus/Consensus.sol [hunk 3]
```diff
@@ -91,29 +103,28 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
...
     /** Assigned core for a given chainId */
     mapping(bytes20 /* chainId */ => address /* core */) public assignments;
 
-    /** Precommitts from cores for metablockchains. */
-    mapping(address /* core */ => Precommit) public precommits;
+    /** Core statuses. */
+    mapping(address /* core */ => CoreStatus /* coreStatus */) public coreStatuses;
+
+    /** Linked-list of committees. */
+    mapping(address => address) public committees;
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol]:
```
mapping(bytes32 /* metachainId */ => CommitteeI) public committees;
```

Comment (benjaminbollen) [reply contracts/consensus/Consensus.sol]: that way, there is less state build-up on `Consensus`; when a new metablock is precommitted; this mapping can be overwritten.

In return the pruning of `Committees` can be done from the committee itself; it can check whether it can be selfdestructed by checking open balances, and whether it has reported the decision / latest, if the metablockchain has moved on to another blockheight

Comment (pgev) [reply contracts/consensus/Consensus.sol]: We can create a follow up ticket for this. What are your thoughts?

**Review by benjaminbollen — COMMENTED** [review]
Hi ! last night i drafted the storage layout I had in mind for `Consensus.sol` after this ticket. you can find it here

https://github.com/mosaicdao/mosaic-pm/blob/master/specification-discussions/20191206-consensus.md#contract-summary

I can also give it a try to implement the functions according to this storage.  Please evaluate from your side whether this is a sensible proposal cc @deepesh-kn @pgev

File: contracts/consensus/Consensus.sol [hunk 4]
```diff
@@ -317,144 +301,232 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
-            "Core has not precommitted."
+            currentMetablock.round == MetablockRound.Precommitted,
+            "Assigned core must have precommitted a metablock to form a committee."
         );
 
-        require(
-            block.number > precommit.committeeFormationBlockHeight,
-            "Block height must be higher than set committee formation height."
+        uint256 committeeFormationBlockHeight = currentMetablock.roundBlockNumber.add(
+            COMMITTEE_FORMATION_LENGTH
```
Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol]:
```suggestion
            COMMITTEE_FORMATION_DELAY
```

File: contracts/consensus/Consensus.sol [hunk 5]
```diff
@@ -317,144 +301,232 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
-        for (uint256 i = 0; i < COMMITTEE_FORMATION_LENGTH; i = i.add(1)) {
-            seedGenerator[i] = blockhash(segmentHeight);
-            segmentHeight = segmentHeight.sub(1);
-        }
-
-        bytes32 seed = keccak256(
-            abi.encodePacked(seedGenerator)
+        bytes32 seed = hashBlockSegment(
+            currentMetablock.roundBlockNumber,
+            committeeFormationBlockHeight
```
Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol:408]:
```suggestion
    block.number.add(COMMITTEE_FORMATION_LENGTH)
```

File: contracts/consensus/Consensus.sol [hunk 6]
```diff
@@ -317,144 +301,232 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
-     * @param _committeeAddress Committee address that validator wants to enter.
+     * @param _metachainId Metachain id for the committee to enter.
      * @param _validator Validator address to enter.
      * @param _furtherMember Validator address that is further member
      *                       compared to the `_validator` address
      */
     function enterCommittee(
-        address _committeeAddress,
+        bytes32 _metachainId,
+        CoreI _core,
```
Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol:438]: Please confirm `enterCommittee` changes is not part of this PR right? 
as per https://github.com/mosaicdao/mosaic-pm/blob/master/specification-discussions/20191206-consensus.md

In that case, we will create an issue for that.

Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol:438]: Why is `_core` needed in the parameter?
We can get `core` from `metachainId`.

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
👍 
Lot of changes. Added a few comments.

File: contracts/consensus/Consensus.sol (hunk 4 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol]: Ah, nice catch. I've not noticed that there are two different variables for mixing length and formation delay. Fixed.

File: contracts/consensus/Consensus.sol (hunk 5 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol:408]: Done.

File: contracts/consensus/Consensus.sol (hunk 6 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol:438]: Confirmed. We should create a new ticket for that one to be merged after this one is merged.

Comment (pgev) [reply contracts/consensus/Consensus.sol:438]: We need to check that validator is active in `some` core.
We need metachain id to identify the metablockchain for which committee has been created and the validator is joining.
In the model of shared security, committee will contain validators from different cores.

File: contracts/consensus/Consensus.sol [hunk 7]
```diff
@@ -1178,32 +1178,33 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
+        uint256 end,
+        uint256 length
     )
         private
         view
         returns (bytes32 seed_)
     {
+        assert(length > 0);
+
+        uint256 begin = end.add(uint256(1)).sub(length);
```
Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol]: 1. `hashBlockSegment` is used only at one place, so we need to pass the length, its already a constant.
2. For me, It was difficult to understand this part of the code, just a suggestion.
Instead of adding `add(uint256(1))` here, can be do the following in the for loop: 

```
for (uint256 i = 1; i <= length; i = i.add(1)) {
            seedGenerator[i] = blockhash(begin.add(i));
        }	        
```

File: contracts/consensus/Consensus.sol (hunk 6 above)
Comment (deepesh-kn) [reply contracts/consensus/Consensus.sol:438]: Created a ticket for this https://github.com/mosaicdao/mosaic-1/issues/147

Comment (deepesh-kn) [reply contracts/consensus/Consensus.sol:438]: This will be updated in https://github.com/mosaicdao/mosaic-1/issues/147, so resolving this.

**Review by deepesh-kn — DISMISSED** [review]
looks good. I added a comment and have created a followup ticket https://github.com/mosaicdao/mosaic-1/issues/147 to update `enterCommittee`

File: contracts/committee/Committee.sol [hunk 8]
```diff
@@ -272,7 +283,6 @@ contract Committee is MasterCopyNonUpgradable, ConsensusModule, CommitteeI {
 
         proposal = _proposal;
 
-        // @qn (pro): In case of 7 quorum should be 4 or 5.
         quorum = _committeeSize * COMMITTEE_SUPER_MAJORITY_NUMERATOR /
         COMMITTEE_SUPER_MAJORITY_DENOMINATOR;
```
Comment (benjaminbollen) [inline contracts/committee/Committee.sol]: missing indentation

File: contracts/consensus/Consensus.sol [hunk 9]
```diff
@@ -317,144 +301,232 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
-        for (uint256 i = 0; i < COMMITTEE_FORMATION_LENGTH; i = i.add(1)) {
-            seedGenerator[i] = blockhash(segmentHeight);
-            segmentHeight = segmentHeight.sub(1);
-        }
-
-        bytes32 seed = keccak256(
-            abi.encodePacked(seedGenerator)
+        bytes32 seed = hashBlockSegment(
+            committeeFormationBlockHeight,
+            COMMITTEE_FORMATION_LENGTH
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol]: why are we passing the constant to an internal function?

consider instead passing the `metablockHash`, because otherwise if multiple precommits happen in the same Ethereum block, they will all have the same seed, and thus the same committee. 

cc @deepesh-kn @jayeshbairagi can we make this a separate task?
```
        bytes32 seed = hashBlockSegments(
            currentMetablock.metablockHash,
            committeeFormationHeight);
```
(reference https://github.com/mosaicdao/mosaic-pm/blob/master/specification-discussions/20191206-consensus.md)

File: contracts/consensus/Consensus.sol [hunk 10]
```diff
@@ -317,144 +301,232 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
+            _validator != address(0),
+            "Validator's address is 0."
+        );
+
+        require(
+            _furtherMember != address(0),
+            "Further validator's address is 0."
+        );
+
+        require(
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol:464]: can we collapse these into a modifier 
```
onlyAssignedCore(_metachainId)
```
in a new ticket?

File: contracts/consensus/Consensus.sol [hunk 11]
```diff
@@ -508,15 +580,17 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
             "Block header does not match with vote message source."
         );
 
-        // Makes sure that assigned core is active.
-        address core = assignments[_metachainId];
+        bytes32 metablockHash = goToCommittedRound(_metachainId);
+
+        CoreI core = assignments[_metachainId];
         require(
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol:587]: ok for now, but it's more of an assert than anything; the assigned core should be active

File: contracts/consensus/Consensus.sol [hunk 12]
```diff
@@ -593,31 +668,34 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
-            core != address(0),
+            core != CoreI(0),
             "Core does not exist for given metachain."
         );
 
         // Specified core must have creation lifetime status.
         require(
-            coreLifetimes[core] == CoreLifetime.creation,
+            coreLifetimes[address(core)] == CoreLifetime.creation,
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol:681]: the mapping should probably be from `CoreI => CoreLifetime`?

File: contracts/consensus/Consensus.sol [hunk 13]
```diff
@@ -662,7 +740,8 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
             "Core lifetime status must be genesis or active."
         );
 
-        CoreI(_core).logout(msg.sender);
+        _core.logout(msg.sender);
+
         reputation.deregister(msg.sender);
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol:744]: @deepesh-kn @jayeshbairagi can we make a ticket to make sure that (co)consensus *first* calls reputation, then the core? not the other way around, thx

File: contracts/consensus/Consensus.sol [hunk 14]
```diff
@@ -1058,16 +1166,49 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
+        private
+        view
+        returns (bytes32 seed_)
+    {
+        assert(length > 0);
+
+        uint256 begin = end.add(uint256(1)).sub(length);
+
+        require(
+            block.number >= end && block.number < begin.add(uint256(256)),
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol]: `block.number` must be strictly greater than `end`, bc the blockhash of the current block is not yet known

File: contracts/consensus/Consensus.sol (hunk 7 above)
Comment (benjaminbollen) [reply contracts/consensus/Consensus.sol]: 1. assuming you meant "we don't" need to pass the length? agreeing with this comment

File: contracts/consensus/Consensus.sol [hunk 15]
```diff
@@ -1058,16 +1166,49 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
+            "Blocksegment is not in the most recent 256 blocks."
         );
 
-        committee_ = CommitteeI(committeeAddress);
+        bytes32[] memory seedGenerator = new bytes32[](length);
+        for (uint256 i = 0; i < length; i = i.add(1)) {
+            seedGenerator[i] = blockhash(begin.add(i));
+        }
+
+        seed_ = keccak256(
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol:1207]: repeating comment made above; in this seed we want to hash the `metablockHash` as well; can be a separate task

**Review by benjaminbollen — CHANGES_REQUESTED** [review]
Looks very good, great progress; thanks, and sorry for the long road to get here;

mostly requesting a small change, can you look and assert this point?
https://github.com/mosaicdao/mosaic-1/pull/103/files#r359458898

the other comments should largely be captured in new tickets; can you assist here @jayeshbairagi we can go over them tomorrow

File: contracts/consensus/Consensus.sol (hunk 7 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol]: @deepesh-kn with regards to the 2. point I would rather keep existing implementation.
@deepesh-kn @benjaminbollen wrt 1. point, I am not sure why should not be function written as a standalone function. However, I will change it as you both have proposed.

File: contracts/committee/Committee.sol (hunk 8 above)
Comment (pgev) [reply contracts/committee/Committee.sol]: Done.

File: contracts/consensus/Consensus.sol (hunk 9 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol]: Done.

File: contracts/consensus/Consensus.sol (hunk 10 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol:464]: In the current implementation of `enterCommittee` the core of the validator is specified by its address (not metablockchain id). In upcoming implementation of `enterCommittee` (according to the newest proposal) we can.

File: contracts/consensus/Consensus.sol (hunk 11 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol:587]: @benjaminbollen do you mean here, that cores in `assignments` storage mapping will not keep halted/corrupted cores?

File: contracts/consensus/Consensus.sol (hunk 12 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol:681]: To my knowledge, solidity mapping key can be only elementary type.

Comment (pgev) [reply contracts/consensus/Consensus.sol:681]: Mapping types use the syntax mapping(_KeyType => _ValueType) and variables of mapping type are declared using the syntax mapping(_KeyType => _ValueType) _VariableName. The _KeyType can be any built-in value type plus bytes and string. User-defined or complex types such as contract types, enums, mappings, structs or array types apart from bytes and string are not allowed. _ValueType can be any type, including mappings, arrays and structs.

https://solidity.readthedocs.io/en/v0.6.0/types.html#mapping-types

File: contracts/consensus/Consensus.sol (hunk 7 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol]: Done.

File: contracts/consensus/Consensus.sol (hunk 14 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol]: Yes. 🤦. Thanks!

File: contracts/consensus/Consensus.sol (hunk 15 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol:1207]: Done.

File: contracts/consensus/Consensus.sol (hunk 7 above)
Comment (deepesh-kn) [reply contracts/consensus/Consensus.sol]: > 1. assuming you meant "we don't" need to pass the length? agreeing with this comment

yes, I meant `we don't`.

**Review by benjaminbollen — APPROVED** [review]
LGTM!
